packages/flutter/lib/src/widgets/spell_check.dart

```diff
@@ -0,0 +1,322 @@
+            TextRange(
+                start: currentSpan.range.start + offset, end: searchStart),
+            currentSpan.suggestions
+        );
+        correctedSpellCheckResults.add(adjustedSpan);
+      } else {
+        regex = RegExp('\\b$currentSpanText\\b');
+        foundIndex = newText.substring(searchStart).indexOf(regex);
+
+        if (foundIndex >= 0) {
+          foundIndex += searchStart;
+          spanLength = currentSpan.range.end - currentSpan.range.start;
+          searchStart = foundIndex + spanLength;
+          adjustedSpan = SuggestionSpan(
+              TextRange(start: foundIndex, end: searchStart),
+              currentSpan.suggestions
+          );
+          offset = foundIndex - currentSpan.range.start;
+
+          correctedSpellCheckResults.add(adjustedSpan);
+        }
+      }
+      spanPointer++;
+    }
+
+    return correctedSpellCheckResults;
+  }
+
+  @override
+  TextSpan buildTextSpanWithSpellCheckSuggestions(
+      TextEditingValue value,
+      bool composingWithinCurrentTextRange,
+      TextStyle? style,
+      TextStyle misspelledTextStyle,
+      SpellCheckResults spellCheckResults) {
+    List<SuggestionSpan>? correctedSpellCheckResults;
+
+    final List<SuggestionSpan> rawSpellCheckResults =
+        spellCheckResults.suggestionSpans;
+    final String spellCheckResultsText = spellCheckResults.spellCheckedText;
+
+    if (spellCheckResultsText != value.text) {
+      correctedSpellCheckResults = correctSpellCheckResults(
+          value.text, spellCheckResultsText, rawSpellCheckResults);
+    } else {
+      correctedSpellCheckResults = rawSpellCheckResults;
+    }
+
+    return TextSpan(
+        style: style,
+        children: buildSubtreesWithMisspelledWordsIndicated(
+            correctedSpellCheckResults,
+            value,
+            style,
+            misspelledTextStyle,
+            composingWithinCurrentTextRange
+        )
+      );
+  }
+
+  /// Builds [TextSpan] subtree for text with misspelled words.
+  static List<TextSpan> buildSubtreesWithMisspelledWordsIndicated(
+      List<SuggestionSpan>? spellCheckSuggestions,
+      TextEditingValue value,
+      TextStyle? style,
+      TextStyle misspelledStyle,
+      bool composingWithinCurrentTextRange) {
+    final List<TextSpan> tsTreeChildren = <TextSpan>[];
+
+    int textPointer = 0;
+    int currSpanPointer = 0;
+    int endIndex;
+    SuggestionSpan currSpan;
+    final String text = value.text;
+    final TextRange composingRegion = value.composing;
+    final TextStyle composingTextStyle =
+        style?.merge(const TextStyle(decoration: TextDecoration.underline)) ??
+            const TextStyle(decoration: TextDecoration.underline);
+    final TextStyle misspelledJointStyle =
+        style?.merge(misspelledStyle) ?? misspelledStyle;
+    bool textPointerWithinComposingRegion = false;
+    bool currSpanIsComposingRegion = false;
+
+    while (spellCheckSuggestions != null &&
```

Comment:
spellCheckSuggestions is not updating, we don't need to check this on each iteration of the loop. We can make this as a if check:
```dart
if (spellCheckSuggestions != null) {
  while()..
}```